anchor/src/setup.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use failure::bail;
+use std::{io::Write, path::Path, process::Command};
+
+enum PackageManager {
+    AptGet,
+    Pacman,
+    Yum,
+    Brew,
+}
+
+impl PackageManager {
+    fn get_system_package_manager() -> Result<PackageManager, failure::Error> {
+        let pkg_mgr = if cfg!(target_os = "linux") {
+            if is_bin_on_path("apt-get")? {
+                PackageManager::AptGet
+            } else if is_bin_on_path("pacman")? {
+                PackageManager::Pacman
+            } else if is_bin_on_path("yum")? {
+                PackageManager::Yum
+            } else {
+                bail!("Unable to find supported package manager (yum, apt-get, or pacman)");
+            }
+        } else if cfg!(target_os = "macos") {
+            if is_bin_on_path("brew")? {
+                PackageManager::Brew
+            } else {
+                bail!("Missing package manager Homebrew (https://brew.sh/)");
+            }
+        } else {
+            bail!("Unknown OS");
+        };
+
+        Ok(pkg_mgr)
+    }
+
+    fn apt_get<P: AsRef<Path>>(bin: P) -> Result<(), failure::Error> {
+        let status = Command::new("sudo")
+            .arg("apt-get")
+            .arg("install")
+            .arg(bin.as_ref())
+            .arg("-y")
+            .status()?;
+
+        if !status.success() {
+            bail!("Running apt-get failed!");
+        }
+
+        Ok(())
+    }
+
+    fn pacman<P: AsRef<Path>>(bin: P) -> Result<(), failure::Error> {
+        let status = Command::new("sudo")
+            .arg("pacman")
+            .arg("-Syu")
+            .arg(bin.as_ref())
+            .arg("--noconfirm")
+            .status()?;
+
+        if !status.success() {
+            bail!("Running pacman failed!");
+        }
+
+        Ok(())
+    }
+
+    fn yum<P: AsRef<Path>>(bin: P) -> Result<(), failure::Error> {
+        let status = Command::new("sudo")
+            .arg("yum")
+            .arg("install")
+            .arg(bin.as_ref())
+            .arg("-y")
+            .status()?;
+
+        if !status.success() {
+            bail!("Running yum failed!");
+        }
+
+        Ok(())
+    }
+
+    fn brew<P: AsRef<Path>>(bin: P) -> Result<(), failure::Error> {
+        let status = Command::new("brew")
+            .arg("install")
+            .arg(bin.as_ref())
+            .status()?;
+
+        if !status.success() {
+            bail!("Running brew failed!");
+        }
+
+        Ok(())
+    }
+
+    fn install_binary<P: AsRef<Path>>(&self, bin: P) -> Result<(), failure::Error> {
+        match self {
+            PackageManager::AptGet => Self::apt_get(bin),
+            PackageManager::Pacman => Self::pacman(bin),
+            PackageManager::Yum => Self::yum(bin),
+            PackageManager::Brew => Self::brew(bin),
+        }
+    }
+
+    fn install_cmake(&self) -> Result<(), failure::Error> {
```

Comment:
Might be worth making this a `Result<T>` where `T` is a sum type of installed and already-present.